### DISCUSSION TITLE
Update dependencies to resolve security vulnerabilities in sub-dependencies

### DISCUSSION THREAD

**Discussion by mattpardee** [renovatebot/renovate#22049] [Suggest an Idea]
**What would you like Renovate to be able to do?**
GitHub informs of vulnerabilities for sub-dependencies present in package-lock.json, but Renovate doesn't cut PRs for updating the top-level dependencies that would resolve those vulnerabilities.

**Describe the solution you'd like**
For Renovate to know when there's a vulnerability in deps' sub-dependencies and then cut a PR for updating the top-level deps to the version that contains the resolution of the security issue.

**Describe alternatives you've considered**
A more regular schedule to update our top-level dependencies (e.g. every day).

**Additional context**
We discussed this [on Twitter](https://twitter.com/rarkins/status/1085570753112694784). Thanks Rhys!

**Comment by rarkins** [comment]
Thank you for making an issue for this to track progress. 

When Renovate reads the vulnerability alerts, it converts them into packageRules. However the `depName` for each rule will be the vulnerable dependency and not the top level dependency itself. 

Two possibilities: 

1. The top level dependency needs updating and the vulnerable transitive dependency will get updated at the same time (easy)

2. The top level dependency is up to date but the lock file holding back the vulnerable transitive dependency.  (Harder)

For point 1, we need to trace back the vulnerable dependency to its top level and make sure they one is updated like usual. 

For point 2, we need to poison the lock file to force the update of the vulnerable transitive dependency. The question is how to describe this. 

Another possibility is to trigger lock file maintenance of the whole file if we detest a transitive vulnerability that can’t be solved at the top level.

**Comment by songkeys** [comment]
@rarkins 

I believe the second point is the more reasonable, which is dependabot's current behavior.

> The question is how to describe this.

What do you mean by this? How to describe the API field or how to implement it?

**Comment by rarkins** [comment]
This is now implemented for npm@6 behind the feature flag `transitiveRemediation`. I've flipped it on for all public repos in the app on github.com and it's found thousands of additional vulnerabilities it's been able to remediate, so it seems like a success. I'll add npm@7 support and then close this issue. We can/should have individual issues for others with lock files like yarn, bundler, composer.

**Comment by cameron-martin** [comment]
That's awesome. Does this work for yarn too, or only npm?

**Comment by viceice** [comment]
Currently only npm < v7

**Comment by kunalnagar** [comment]
@rarkins @viceice I have the `transitiveRemediation` flag turned on. I get the following in the Dependency Dashboard but no PRs have actually been created 🤔 

```
{
  "renovateVersion": "25.18.8"
}
```

Happy to provide more info if you'd like!

![image](https://user-images.githubusercontent.com/2741371/118031640-c3d32c80-b334-11eb-985f-9491d742194a.png)

**Comment by rarkins** [comment]
Please provide a reproduction in a public repo so we can take a look.

**Comment by kunalnagar** [comment]
I created [this repo](https://github.com/kunalnagar/test-renovate-cve) but I've been having issues with Renovate trying to generate a Dependency Dashboard. I see the logs have failed with an `unknown-error`, not sure what that means.

I do have Renovate enabled for this repo as a GitHub app and I've enabled Dependabot Alerts and Dependency graph.

**Edit:** Pasting screenshot of Dependabot Vulnerability in `trim`

![image](https://user-images.githubusercontent.com/2741371/118037129-71e1d500-b33b-11eb-8589-c5cb7f199de7.png)

**Comment by rarkins** [comment]
But is transitive remediation enabled in that repo?

**Comment by kunalnagar** [comment]
It was, [until I removed it here](https://github.com/kunalnagar/test-renovate-cve/commit/13b9bf1ef140b76eb443931dd8f6daafb0ae45b4) and suddenly started seeing alerts show up.

I can't explain why that happened 🤔 

Edit: I noticed in [this comment](https://github.com/renovatebot/renovate/issues/3080#issuecomment-792643886) where you say that you've flipped it for all public repos, so maybe that's why?

**Comment by rarkins** [comment]
> Edit: I noticed in [this comment](https://github.com/renovatebot/renovate/issues/3080#issuecomment-792643886) where you say that you've flipped it for all public repos, so maybe that's why?

You're right, I forgot :)

**Comment by rarkins** [comment]
It would seem like `trim` is the only one in the reproduction repo which can't be remediated?

**Comment by kunalnagar** [comment]
> It would seem like trim is the only one in the reproduction repo which can't be remediated?

Yep. 

Do you have any thoughts on what the `unkown-error` was in the logs before I removed the `transitiveRemediation` flag? If it comes up again, do you have any possible next steps to debug?

**Comment by rarkins** [comment]
I'm not sure re: the unknown error, but please try to capture/save the full log in future if it happens again. Meanwhile the `trim` dashboard problem is described in #9977

**Comment by kunalnagar** [comment]
> I'm not sure re: the unknown error, but please try to capture/save the full log in future if it happens again.

Here are the logs from the Renovate Dashboard:

<details>
<summary>Logs</summary>

```
INFO: Repository started
{
  "renovateVersion": "25.18.8"
}
DEBUG: Using localDir: /mnt/renovate/gh/kunalnagar/test-renovate-cve
DEBUG: Repository cache not found
{
  "cacheFileName": "/tmp/renovate-cache/renovate/repository/github/kunalnagar/test-renovate-cve.json"
}
ERROR: Repository has unknown error
{
  "err": {
    "errno": -13,
    "code": "EACCES",
    "syscall": "mkdir",
    "path": "/tmp/renovate-cache/__renovate-private-cache",
    "message": "EACCES: permission denied, mkdir '/tmp/renovate-cache/__renovate-private-cache'",
    "stack": "Error: EACCES: permission denied, mkdir '/tmp/renovate-cache/__renovate-private-cache'"
  }
}
DEBUG: Unknown res
{
  "res": "unknown-error"
}
DEBUG: Repository timing splits (milliseconds)
{
  "splits": {},
  "total": 42
}
INFO: Repository finished
{
  "durationMs": 42
}
```

</details>

> Meanwhile the trim dashboard problem is described in #9977

Thanks, I'll keep an eye on it 🙂

**Comment by rarkins** [comment]
Oh sorry - that was a temporary error today when unfortunately a particular worker somehow got a cache permissions issue. It was terminated and hopefully it does not happen again (it hasn't happened before at least..). I will mark all these comments as resolved

**Comment by kunalnagar** [comment]
@rarkins I just wanted to check if there's a rough timeline on `yarn` support for the `transitiveRemediation` flag? I couldn't find any active PRs around this at the moment 🤔

**Comment by The-Code-Monkey** [comment]
Do we have a timeframe for updating the yarn lock file dependencies for security reasons?

**Comment by viceice** [comment]
Idea:

Run a forced lockfile maintenance for managers which support it and has transitive dependency security issues?

@rarkins WDYT?

**Comment by rarkins** [comment]
That could often fix transitive vulnerabilities although not always. Anyway, can be useful.

**Comment by milind009** [comment]
Hi @rarkins , does renovate have the capability to update specific version of a dependency in case there are multiple versions for a dependency present in the lockfile. For e.g a repo uses 3 versions for dependency A in its lockfile -> 1.0, 1.1, 1.5. Now there is a vulnerability discovered with dependency A for versions <= 1.1 with secure version being >= 1.5. Will renovate be able to resolve the security vulnerability associated with A in this repo's lockfile and create a PR for it assuming that the current semver constraints allows update of dependency A -> 1.5

**Comment by rarkins** [comment]
The current approach is supported for npm only, and you are using versions which aren't valid npm versions, so I don't know if we're on the same page or not. The answer is really "maybe". For example I'm not sure that GitHub's alerts would even correctly pick up the situation you're describing anyway, or if they'd just alert for 1.0 or 1.1. If you can reproduce it with an example, I can answer you with more certainty

**Comment by khvn26** [comment]
Solution for uv: https://github.com/renovatebot/renovate/pull/44262

I focused on uv since it's our primary use case at Flagsmith. I've managed to surface uv's lockfile-only deps so `osvVulnerabilityAlerts` can match them, and remediate each via `uv lock --upgrade-package` which bumps a single locked package in place. It seems that the approach will generalise well to other ecosystems supporting targeted lock bumps.
